Doc/library/tkinter.rst

```diff
@@ -91,54 +91,27 @@ Python interpreter.
-the interpreter runs an event loop (i.e. :func:`Tk.mainloop`) and responds to events.
-Because it is single-threaded, event handlers must respond quickly, otherwise they
-will block other events from being processed. To avoid this, any long-running
-computations should not run in an event handler, but are either broken into smaller
-pieces using timers, or run in another thread. This is different from many GUI
-toolkits where the GUI runs in a completely separate thread from all application
-code including event handlers.
-
-If the Tcl interpreter is not running the event loop and processing events, any
-:mod:`tkinter` calls made from threads other than the one running the Tcl
-interpreter will fail.
-
-A number of special cases exist:
-
-  * Tcl/Tk libraries can be built so they are not thread-aware. In this case,
-    :mod:`tkinter` calls the library from the originating Python thread, even
-    if this is different than the thread that created the Tcl interpreter. A global
-    lock ensures only one call occurs at a time.
-
-  * While :mod:`tkinter` allows you to create more than one instance of a :class:`Tk`
-    object (with it's own interpreter), all interpreters that are part of the same
-    thread share a common event queue, which gets ugly fast. In practice, don't create
-    more than one instance of :class:`Tk` at a time. Otherwise, it's best to create
-    them in separate threads and ensure you're running a thread-aware Tcl/Tk build.
-
-  * Blocking event handlers are not the only way to prevent the Tcl interpreter from
-    reentering the event loop. It is even possible to run multiple nested event loops
-    or abandon the event loop entirely. If you're doing anything tricky when it comes
-    to events or threads, be aware of these possibilities.
-
-  * There are a few select :mod:`tkinter` functions that presently work only when
-    called from the thread that created the Tcl interpreter.
+Tkinter strives to allow any calls to its API from any Python threads, without any limits, as expected from a Python module. Due to Tcl's architectural restrictions, however, that stem from its vastly different threading model, this is not always possible.
+
+Tcl's execution model is based on cooperative multitasking. Control is passed between multiple interpreter instances by sending events (see `event-oriented programming -- Tcl/Tk wiki <https://wiki.tcl.tk/1772>`_ for details).
+
+A Tcl interpreter instance has only one stream of execution and, unlike many other GUI toolkits, Tcl/Tk doesn't provide a blocking event loop. Instead, Tcl code is supposed to pump the event queue by hand at strategic
+moments (save for events that are generated explicitly in the same OS thread -- these are handled immediately by simply passing control from sender to the handler). As such, all Tcl commands are designed to work without an event loop running -- only the event handlers will not fire until the queue is processed.
+
```

Comment:
For our purposes, does "Tcl/Tk doesn't provide a blocking event loop" really mean that Tcl/Tk does not guarantee that events will be continuously processed? Why include the parts about pumping the event queue (which I assume means calling Tcl_DoOneEvent)? The "designed to work without an event loop" also seems extraneous. Why do we care?

Reply:
Or to put it another way, is the key consideration that the interpreter is "actively processing events"? (to use a technically inaccurate but common sense phrase)

Reply:
> For our purposes, does "Tcl/Tk doesn't provide a blocking event loop" really mean that Tcl/Tk does not guarantee that events will be continuously processed? 

The multithreaded GUI model that Tkinter implements typically requires that only the UI thread can access GUI elements. As such, any operation with GUI whatsoever is normally expected to require an event loop running.

This is not the case with Tcl/Tk and thus Tkinter. Tcl/Tk commands _are_ operational without an event loop to some extent. _But no-one really knows what that extent is, and how to compensate._ (Look e,g. at https://stackoverflow.com/questions/4083796/how-do-i-run-unittest-on-a-tkinter-app and http://code.activestate.com/recipes/578964-robust-unittesting-of-tkinter-menu-items-with-mock/ (the latter is linked to from a deleted answer in the former that you probably can't see) to see the sheer extent of the weird things that people tried and failed to get something resembling a predictable result!)